### PR TITLE
Specify device in DataCollator 

### PR DESCRIPTION
**Pull request by mfuntowicz**
By setting `device` parameter in `DataCollator` we're able to allocate tensors directly on the right device at tensor creation and avoid moving data around afterwards.

This effectively avoid some snippet like this in the Trainer: 

```python

for k, v in some_dict.items():
     v.to(device)
```

### CODE REVIEW

**Comment by BramVanroy** [comment]
@mfuntowicz Perhaps it's useful to pull this through to the encode methods of the tokenizers so that you can pass a device and if return_tensors is used, the tensor are automatically pushed to the correct device?

**Comment by mfuntowicz** [comment]
I tend to agree we need to have a device argument on the tokenizer `encode_like` methods, that would remove the need to iterate over the items to relocate on GPU/TPU.

Is there a common scheme we can use to do this on both Pytorch & TensorFlow (_I'm not very familiar with TensorFlow_)? 

I suspect strings might be the easiest way to handle this: 

- PyTorch: `device="cpu"`  || `device="cuda:0"`
- TensorFlow: `device=`/device:cpu` || `/device:gpu:0` 

I need to test things out futher here to see what the API may looks like! 👍 
Let's see what the other think of the propal for the Trainer and we can follow up on a dedicated PR asap.

**Comment by BramVanroy** [comment]
> I tend to agree we need to have a device argument on the tokenizer `encode_like` methods, that would remove the need to iterate over the items to relocate on GPU/TPU.
> 
> Is there a common scheme we can use to do this on both Pytorch & TensorFlow (_I'm not very familiar with TensorFlow_)?
> 
> I suspect strings might be the easiest way to handle this:
> 
> * PyTorch: `device="cpu"`  || `device="cuda:0"`
> * TensorFlow: `device="/device:cpu"`  ||  `device="/device:gpu:0"`
> 
> I need to test things out futher here to see what the API may looks like! 👍
> Let's see what the other think of the propal for the Trainer and we can follow up on a dedicated PR asap.

Sorry, I also don't have much experience with TF so I can't really chip in on that. I guess the encode methods can accept a `device=` property that can be of type `Union[str, int, tf.device, torch.device]`. The device constructor (pt or tf) can depend on the already available `return_tensors` part which can be None, pt, or tf.

- `str`: use it to initialize the `torch.device(str)` or `tf.device(str))` - allows the users with a lot of freedom in case they  want do something like `tf.device('/job:bar/task:0/device:gpu:2')` (example from [docs](https://www.tensorflow.org/api_docs/python/tf/device))
- `int`: assume this is a GPU device id: `torch.device(f"cuda:{int}")` or `torch.device(f"/device:gpu:{int}")`
- a `device`: use the device

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -49,20 +57,23 @@ def collate_batch(self, features: List[InputDataClass]) -> Dict[str, torch.Tenso
         # on the whole batch.
         first = features[0]
 
+        # Cache to avoid class lookup
+        device = self.device
```
Comment (julien-c) [inline src/transformers/data/data_collator.py:61]: is this optimization really worth it?

**Comment by julien-c** [comment]
@mfuntowicz is the performance improvement of this change significant? How can we measure it?

**Comment by mfuntowicz** [comment]
Let me run some epochs to report quantitative numbers

**Comment by BramVanroy** [comment]
Perhaps useful discussions here: https://stackoverflow.com/questions/28597014/python-why-is-accessing-instance-attribute-is-slower-than-local

In 100 calls to the variable (self or local) the answer reports 3 seconds difference. Not sure if that is worth it.

I'm all for speed improvements but I'm also really fond of readability.

File: src/transformers/data/data_collator.py (hunk 1 above)
Comment (manishiitg) [reply src/transformers/data/data_collator.py:61]: .to(device) is already being done in train() method in trainer.

FYI

**Comment by BramVanroy** [comment]
cc @LysandreJik

**Comment by BramVanroy** [comment]
Should this still be implemented, or is this PR superseded by another one? @mfuntowicz @julien-c

**Comment by BramVanroy** [comment]
Bump @julien-c  @mfuntowicz

**Comment by BramVanroy** [comment]
Should I keep bumping this? @sgugger

**Comment by sgugger** [comment]
I don't think it's useful: the Trainer handles it already and for people that want to run their custom training loop, Accelerate can handle that.

**Pull request closed without merging**